packages/server/src/utils/request-logger.js

```diff
@@ -0,0 +1,145 @@
+/*
+ * Module inspired by 'koa-logger'. Adapted and extended to our needs.
+ */
+import counterFunc from 'passthrough-counter'
+import bytes from 'bytes'
+import chalk from 'chalk'
+
+export function logRequests({ ignoreUrls } = {}) {
+  return async function middleware(ctx, next) {
+    if (ignoreUrls && ctx.req.url.match(ignoreUrls)) {
+      return next()
+    }
+    // request
+    const start = Date.now()
+
+    logRequest(ctx)
+
+    try {
+      await next()
+    } catch (err) {
+      logResponse({
+        ctx,
+        start,
+        err
+      })
+      throw err
+    }
+
+    // calculate the length of a streaming response
+    // by intercepting the stream with a counter.
+    // only necessary if a content-length header is currently not set.
+    const {
+      body,
+      response: { length }
+    } = ctx
+    let counter
+    if (length === null && body && body.readable) {
+      ctx.body = body.pipe((counter = counterFunc())).on('error', ctx.onerror)
+    }
+
+    // log when the response is finished or closed,
+    // whichever happens first.
+    const { res } = ctx
+
+    const onfinish = done.bind(null, 'finish')
+    const onclose = done.bind(null, 'close')
+
+    res.once('finish', onfinish)
+    res.once('close', onclose)
+
+    function done(event) {
+      res.removeListener('finish', onfinish)
+      res.removeListener('close', onclose)
+      logResponse({
+        ctx,
+        start,
+        length: counter ? counter.length : length,
+        event
+      })
+    }
+  }
+}
+
+function findLogger(ctx) {
+  return ctx.log?.child({ name: 'requests' }) ?? null
+}
+
+function logRequest(ctx) {
+  const log = findLogger(ctx)
+  log?.info(
+    {
+      req: ctx.req
+    },
+    `${chalk.gray('<--')} ${chalk.bold(ctx.method)} ${chalk.gray(
+      ctx.originalUrl
+    )}`
+  )
+}
+
+function logResponse({
+  ctx,
+  start,
+  length,
+  err,
+  event
+}) {
+  const log = findLogger(ctx)
+
+  if (!log) {
+    return
+  }
+
+  // get the status code of the response
+  const status = err ? err.status || 500 : ctx.status || 404
+
+  // set the color of the status code;
+  const s = (status / 100) | 0
+  const color = colorCodes.hasOwnProperty(s) ? colorCodes[s] : colorCodes[0]
+
+  // get the human readable response length
+  const formattedLength = [204, 205, 304].includes(status)
+    ? ''
+    : length == null
+      ? '-'
+      : bytes(length).toLowerCase()
+
+  const upstream = err
+    ? chalk.red('xxx')
+    : event === 'close'
+      ? chalk.yellow('-x-')
+      : chalk.gray('-->')
+
+  const formattedTime = formatTime(start)
+
+  const mergingObject = {
```

Comment:
This variable name confuses me. What's `mergingObject` about?

Reply:
It's pino language :)

Reply:
Ah, i see :D

Reply:
name is bad though, changing